[sw/dif] Introduce a library for manipulating volatile MMIO memory

silvestrst opened the pull request:
This change has been factored out of the @mcy #1135 SPI DIF implementation. This API is expected to be used by all peripheral DIF implementations, and is encouraged to be used in any DIF consumers that might require HW access outside of the capabilities provided by the DIFs.

Related pull requests:
#1135 SPI DIF
#1111 UART DIF

Related open issues:
#1076 UART DIF

mcy commented:
Actually, what do you think about putting this library under lib/base instead? See https://github.com/lowRISC/opentitan/pull/1190.

silvestrst commented:
@mcy I think that is fine, we had a some discussion about it with @lenary . I think it makes sense to have register access library under libbase.

On the other side, now when I think about it, probably, only DIFs will ever need to access the memory mapped registers, so it also makes sense to keep them under DIF directory. Then again, DIFs would probably need to use libbase :)

I don't have a strong opinion on this, happy to move it in libbase if you guys think that it's a better alternative?

silvestrst commented:
> LGTM. Thanks for factoring this out for me!

No problem

gkelly commented:
> @mcy I think that is fine, we had a some discussion about it with @lenary . I think it makes sense to have register access library under libbase.
> 
> On the other side, now when I think about it, probably, only DIFs will ever need to access the memory mapped registers, so it also makes sense to keep them under DIF directory. Then again, DIFs would probably need to use libbase :)
> 
> I don't have a strong opinion on this, happy to move it in libbase if you guys think that it's a better alternative?

I think the more self-contained this is the better.  If someone wants to build a DIF library I feel like they shouldn't need to get libbase involved.

lenary commented:
Let's put it under libbase, but keep a small meson dependency so DIFs can depend on only register.c/.h rather than having to pull in all of libbase.

gkelly commented:
> Let's put it under libbase, but keep a small meson dependency so DIFs can depend on only register.c/.h rather than having to pull in all of libbase.

That works for me!

silvestrst commented:
Changes in the latest push:
* added @return in read bit function description (was missing)
* added reg32_read_mask function, as sometimes it is useful to read the value of the mask in the register, as opposed to the whole register.

silvestrst commented:
Changes in the latest push:
* changed reg32_read_mask to shift the mask internally (makes more sense, as it needs to shift back the result anyway)
* improved the documentation for reg32_read_mask function

mcy commented:
If `read_mask` is going to take a subword offset, I would make the other mask functions do so, too.

silvestrst commented on sw/device/lib/dif/register.h:
```diff
@@ -0,0 +1,151 @@
+ * @param mask_offset the offset of the mask in the selected register.
+ * @retun return the value of the read mask.
+ */
+inline uint32_t reg32_read_mask(reg32_t base, ptrdiff_t offset, uint32_t mask,
```
This function can be done more efficient - like reg32_get_bit. There is no need for shifting the value twice. Thanks @lenary

mcy replied:
FWIW, all of these functions will be inlined aggressively, so I wouldn't worry about bit tricks too much. Strength reduction will deal with it.

silvestrst commented:
> If `read_mask` is going to take a subword offset, I would make the other mask functions do so, too.

Good point, thanks

mcy commented:
What's missing to land this? I want to poke at some other DIFs and having this in-tree will help.

silvestrst commented:
@mcy In the latest push I have addressed your comments. I think Alex had some comments, but didn't get around to yet to post them. @asb

lenary commented on sw/device/lib/dif/register.c:
```diff
@@ -0,0 +1,17 @@
+extern uint32_t reg32_read(reg32_t, ptrdiff_t);
+extern void reg32_write(reg32_t, ptrdiff_t, uint32_t);
+extern void reg32_clear_mask(reg32_t, ptrdiff_t, uint32_t);
+extern void reg32_set_mask(reg32_t, ptrdiff_t, uint32_t);
```
You need to update these signatures so they match the function definitions.

silvestrst replied:
Oops, will do

mcy commented:
@silvestrst I think it's fine to note that anything beyond read/write is non-atomic and subject to interruption. If a particular register modification needs to be done without interrupts in the way, that should be on the caller. Pay for what you use, as they say.

Also, I hate to do this, but do you think `mmio.h` and `mmio.c` are better names for these files? Mostly because I could imagine wanting to use `register` for something else (like a file containing a struct of 31 words that I could spill all the RISC-V registers into).

silvestrst commented:
> @silvestrst I think it's fine to note that anything beyond read/write is non-atomic and subject to interruption. If a particular register modification needs to be done without interrupts in the way, that should be on the caller. Pay for what you use, as they say.
> 
> Also, I hate to do this, but do you think `mmio.h` and `mmio.c` are better names for these files? Mostly because I could imagine wanting to use `register` for something else (like a file containing a struct of 31 words that I could spill all the RISC-V registers into).

On the subject of atomicity: I actually came to the similar conclusion last evening, however didn't get around to post it. It would be neat if RISC-V had instructions that could facilitate this, I will have a look, but otherwise I think you are right.

On the subject of mmio: mmio seems more descriptive of what the contents do, I will change it, and also put under the libbase.

lenary commented:
> On the subject of atomicity: I actually came to the similar conclusion last evening, however didn't get around to post it. It would be neat if RISC-V had instructions that could facilitate this, I will have a look, but otherwise I think you are right.

RISC-V does not have instructions for doing read-modify-write, except in the A extension, which Ibex does not implement.

I think a reasonable path forward is to document that these functions are not atomic, and may be interrupted, and that it's up to other components to ensure that this does not happen. This should go in the doxygen.

I think the atomicity, or not, of these calls is a DIF issue, and not an issue with this interface. We only expect DIF code to be using this interface, and DIFs do not do interrupt management. If, for example, a DIF were to write to (or read from) several MMIO registers in the same call, this would also not be atomic. In this case, the DIF should probably document which interrupts should be disabled when the DIF is called. Let's pick this discussion back up on the DIFs document.

asb commented:
I think my concern is that volatile and non-atomic operations are hidden behind the API in a non-obvious way. On a system that did implement the A extension, the natural implementation of a number of the operations here would use the atomicrmw ops. If we think there are cases where you want register reading to be atomic and cases where you don't care, we should probably have these via separate helper functions.

Has a more minimal interface been considered - perhaps one that encourages access to device registers via volatile load/store helper functions and relies on separate helpers for manipulating/masking the value. Although convenient, the helpers here potentially encourage a programming style where you repeatedly read a volatile value in a way that might read inconsistent state (e.g. using `reg32_read_mask`). The abstraction here makes me uneasy. I could see an argument that only DIF writers who know what they're doing will be using these functions anyway, but even then it would be good to put more warnings in the doc.

A slightly more meta concern (that's not really in the scope of this PR, but is worth flagging) is regarding atomic accesses to volatile device registers: Have we been adding more device registers where atomic access is important? In cases such like GPIO we provide an interface that allows setting output pins without requiring an atomicrmw. If there are other devices that would benefit from such interfaces we should evaluate adding them, or alternatively implementing support for the A extension in Ibex.

mcy commented:
@asb How would you feel about marking the non-atomic functions as such? e.g. `reg32_nonatomic_read_mask`?

silvestrst commented:
I like @mcy suggestion, @lenary had same idea.

asb commented:
@mcy: I like that idea.

Also, I know this PR is blocking the DIFs work we want to land, so I don't want to block it. I'd be happy for it to land as-is, even happier if we labeled functions explicitly as nonatomic. Even then we might want to revisit some aspects of the API later, so I'm mainly flagging that we probably shouldn't consider this API fully finalised, as there are some issues that probably merit another look once we have more experience with DIF implementation.

silvestrst commented:
> @mcy: I like that idea.
> 
> Even then we might want to revisit some aspects of the API later, so I'm mainly flagging that we probably shouldn't consider this API fully finalised, as there are some issues that probably merit another look once we have more experience with DIF implementation.

I could add this as a NOTE in the commit message as well

silvestrst commented:
In the latest push:
* added "nonatomic" to following functions:
reg32_clear_mask
reg32_set_mask
reg32_clear_bit
reg32_set_bit
*  Added a note to the commit message that this API could be revised in future

I think reg32_read_mask is okay without "nonatomic" in the name - as it only access the register for read once?

asb commented:
@silvestrst sounds sensible to me. I still wonder if having methods that _just_ read/write a 32-bit reg and then separate helpers for masking+shifting (if desired) would be cleaner and more obvious. But happy to go with this style to see how it works out.

silvestrst commented:
@asb something like:
sw_lib_mmio - mmio.h, mmio.c (read, write)
sw_lib_mmio_helpers - mmio_helpers.h, mmio_helpers.c (read_mask, set_mask, get_mask, clear_bit, set_bit) ?

silvestrst commented:
In the last push:
Updated the description of reg32_nonatomic_clear_bit and reg32_nonatomic_set_bit

mcy commented:
++ @silvestrst thanks for pushing this over the finish line!

silvestrst commented:
@mcy thanks
